addons/website_sale_wallet/models/sale_order.py

```diff
@@ -0,0 +1,145 @@
+        self.recompute_wallet_payment()
+        return res
+
+    def recompute_wallet_payment(self):
+        for order in self:
+            if order.order_line_wallet_id:
+                can_be_retained = order.partner_id.current_wallet_id.how_much_can_be_retained(self.get_total_amount(),
+                                                                                              order.currency_id)
+                order.order_line_wallet_id.price_unit = - can_be_retained
+
+    def _get_wallet_product(self):
+        if not self.wallet_product_id:
+            product = self.env['product.product'].search([('default_code', '=', 'PAY_WITH_WALLET')])
+            if product:
+                self.wallet_product_id = product
+            else:
+                self.wallet_product_id = self.env['product.product'].create({
+                    'default_code': 'PAY_WITH_WALLET',
+                    'name': _("Pay with Wallet"),
+                    'type': 'service',
+                    'taxes_id': False,
+                    'supplier_taxes_id': False,
+                    'sale_ok': False,
+                    'purchase_ok': False,
+                    'lst_price': 0,
+                })
+        return self.wallet_product_id
+
+    def _create_gift_cards(self, gift_card_order_line):
+        return [self.env['gift.card'].create({
+            'initial_amount': gift_card_order_line.order_id.currency_id.compute(gift_card_order_line.price_unit,
+                                                                                self.env.company.currency_id),
+            'sale_order_purchase_id': self.id,
+            'website_id': self.env['website'].get_current_website().id
+        }) for _ in range(int(gift_card_order_line.product_uom_qty))]
+
+    def _send_gift_card_mail(self):
+        template = self.env.ref('website_sale_wallet.mail_template_gift_card', raise_if_not_found=False)
+        if template:
+            for order in self:
+                for gift_card in order.generated_gift_card_ids:
+                    order.message_post_with_template(
+                        template.id, composition_mode='comment',
+                        model='gift.card', res_id=gift_card.id,
+                        email_layout_xmlid='mail.mail_notification_light',
+                    )
+
+    def reload_wallet_payment(self):
+        return self.order_line_wallet_id and (
+                self.amount_total == 0
+                or float_compare(self.wallet_balance + self.order_line_wallet_id.price_unit, 0.0, precision_rounding=2) > 0
+        )
+
+
+class SaleOrderLine(models.Model):
+    # Inherited Fields
+    _inherit = "sale.order.line"
+
+    is_wallet_payment = fields.Boolean(default=False)
+    # Relational fields
+    website_id = fields.Many2one("website", related='order_id.website_id')
+
+    # Crud
+    def unlink(self):
+        for order_line in self.filtered(lambda line: line.is_wallet_payment):
```

Comment:
same here, delete cascade shoudl work... if you delete a SO, it delete all Sale Order Line, so you can by cascade delete Wallet_id linked to this order.

Balance = initial_amount - used_amount